fish-rust/src/wutil/gettext.rs

```diff
@@ -1,17 +1,102 @@
+use crate::fish::PACKAGE_NAME;
+use crate::wchar::prelude::*;
+use crate::wchar_ffi::wchar_t;
+use errno::{errno, set_errno};
+use once_cell::sync::{Lazy, OnceCell};
 use widestring::U32CString;
 
-/// Support for wgettext.
+#[cfg(feature = "gettext")]
+mod internal {
+    use libc::c_char;
+    use std::ffi::CStr;
+    extern "C" {
+        fn gettext(msgid: *const c_char) -> *mut c_char;
+        fn bindtextdomain(domainname: *const c_char, dirname: *const c_char) -> *mut c_char;
+        fn textdomain(domainname: *const c_char) -> *mut c_char;
+    }
+    pub fn fish_gettext(msgid: &CStr) -> *const c_char {
+        unsafe { gettext(msgid.as_ptr()) }
+    }
+    pub fn fish_bindtextdomain(domainname: &CStr, dirname: &CStr) -> *mut c_char {
+        unsafe { bindtextdomain(domainname.as_ptr(), dirname.as_ptr()) }
+    }
+    pub fn fish_textdomain(domainname: &CStr) -> *mut c_char {
+        unsafe { textdomain(domainname.as_ptr()) }
+    }
+}
+#[cfg(not(feature = "gettext"))]
+mod internal {
+    use libc::c_char;
+    use std::ffi::CStr;
+    pub fn fish_gettext(msgid: &CStr) -> *const c_char {
+        msgid.as_ptr()
+    }
+    pub fn fish_bindtextdomain(_domainname: &CStr, _dirname: &CStr) -> *mut c_char {
+        std::ptr::null_mut()
+    }
+    pub fn fish_textdomain(_domainname: &CStr) -> *mut c_char {
+        std::ptr::null_mut()
+    }
+}
+
+use internal::*;
+
+// Really init wgettext.
+fn wgettext_really_init() {
+    let package_name = CString::new(PACKAGE_NAME).unwrap();
+    let localedir = CString::new(option_env!("LOCALEDIR").unwrap_or("UNDEFINED")).unwrap();
+    fish_bindtextdomain(&package_name, &localedir);
+    fish_textdomain(&package_name);
+}
+
+fn wgettext_init_if_necessary() {
+    static INIT: OnceCell<()> = OnceCell::new();
+    INIT.get_or_init(wgettext_really_init);
+}
 
 /// Implementation detail for wgettext!.
+/// Wide character wrapper around the gettext function. For historic reasons, unlike the real
+/// gettext function, wgettext takes care of setting the correct domain, etc. using the textdomain
+/// and bindtextdomain functions. This should probably be moved out of wgettext, so that wgettext
+/// will be nothing more than a wrapper around gettext, like all other functions in this file.
 pub fn wgettext_impl_do_not_use_directly(text: &[wchar_t]) -> &'static wstr {
     assert_eq!(text.last(), Some(&0), "should be nul-terminated");
-    let res: *const wchar_t = ffi::wgettext_ptr(text.as_ptr());
-    #[allow(clippy::unnecessary_cast)]
-    let slice = unsafe { std::slice::from_raw_parts(res as *const u32, wcslen(res)) };
-    wstr::from_slice(slice).expect("Invalid UTF-32")
+    // Preserve errno across this since this is often used in printing error messages.
+    let err = errno();
+
+    // TODO don't allocate
+    let key = text.to_vec();
+
+    wgettext_init_if_necessary();
+    #[allow(clippy::type_complexity)]
+    static WGETTEXT_MAP: Lazy<Mutex<HashMap<Vec<wchar_t>, Pin<Box<WString>>>>> =
```

Comment:
This cache needs to be invalidated if we switch locale (e.g. `set -gx LANG foo` in config.fish).

I *think* it would have to be a global for that?

See e.g.

```fish
LANG=de_DE.UTF-8 /usr/bin/fish -c 'set -gx LC_ALL en_US.UTF-8; _ file'
```

Which will correctly print "file" in english. But if you add another `_ file` before the `set`, it will have cached it and print "Datei" both times.